src/library/scala/util/Using.scala

```diff
@@ -0,0 +1,286 @@
+  *                                          which safely wraps resource usage and management in a `Try`.
+  * @define multiResourceSuppressionBehavior If more than one exception is thrown by the operation and closing resources,
+  *                                          the exception thrown ''first'' is thrown, with the other exceptions
+  *                                          [[java.lang.Throwable.addSuppressed(Throwable) added as suppressed exceptions]]
+  *                                          to the one thrown first. This is the case ''unless'' a later exception is
+  *                                          [[scala.util.control.NonFatal fatal]], and the one preceding it is not. In that case,
+  *                                          the first exception is added as a suppressed exception to the fatal one, and the fatal
+  *                                          one is thrown. If an exception is a
+  *                                          [[scala.util.control.ControlThrowable ControlThrowable]], no exception will be added to
+  *                                          it as a suppressed exception.
+  */
+object Using {
+  /** Creates a `Using` from the given resource.
+    *
+    * @note If the resource does not have an implicit [[Resource]] in
+    *       scope, the returned `Using` will be useless.
+    */
+  def apply[R](resource: => R): Using[R] = new Using(resource)
+
+  /** Performs an operation using a resource, and then closes the resource,
+    * even if the operation throws an exception. This method behaves similarly
+    * to Java's try-with-resources.
+    *
+    * $recommendUsing
+    *
+    * If both the operation and closing the resource throw exceptions, the one thrown
+    * when closing the resource is
+    * [[java.lang.Throwable.addSuppressed(Throwable) added as a suppressed exception]]
+    * to the one thrown by the operation, ''unless'' the exception thrown when closing
+    * the resource is [[scala.util.control.NonFatal fatal]], and the one thrown by the
+    * operation is not. In that case, the exception thrown by the operation is added
+    * as a suppressed exception to the one thrown when closing the resource. If an
+    * exception is a [[scala.util.control.ControlThrowable ControlThrowable]], no
+    * exception will be added to it as a suppressed exception.
+    *
+    * @param resource the resource
+    * @param body     the operation to perform with the resource
+    * @tparam R the type of the resource
+    * @tparam A the return type of the operation
+    * @return the result of the operation, if neither the operation nor
+    *         closing the resource throws
+    */
+  def resource[R: Resource, A](resource: R)(body: R => A): A = {
+    @inline def safeAddSuppressed(t: Throwable, suppressed: Throwable): Unit = {
+      // don't `addSuppressed` to something which is a `ControlThrowable`
+      if (!t.isInstanceOf[ControlThrowable]) t.addSuppressed(suppressed)
```

Comment:
You can't safely do this, since cached Exception instances could be thrown. The only safe thing to addSuppressed to is exceptions created by code you control.

Reply:
We need to be able to do this, because silently dropping the exception is a significant loss of functionality. Additionally, it's what Java does, so

- it would be bad to add a feature that kind of does the thing you want it to, but only like halfway (if that) - people will have to use Java instead
- if people call Java code that uses try-with-resources, it will call `addSuppressed` anyway, so hoping that no one calls that method isn't really a safe assumption.

If you're using a cached exception, you can pass in the `Exception`/`Throwable`/`Error` ctor parameter to disable suppression.

We should also honestly make `ControlThrowable` (maybe even `NoStackTrace`) do this by default, but it's a trait, so we can't. In fact, we should also use the ctor parameter for disabling stack traces, but again, trait.

Reply:
@NthPortal Ah, you mean that by specifying a cause Throwable when constructing an Exception suppresses the ability to add suppressed? If so, then I should probably review some cached exceptions.

Reply:
@viktorklang Java (7?) added the following constructor for [`Throwable`](https://docs.oracle.com/javase/8/docs/api/java/lang/Throwable.html#Throwable-java.lang.String-java.lang.Throwable-boolean-boolean-), as well as identical ones for [`Exception`](https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html#Exception-java.lang.String-java.lang.Throwable-boolean-boolean-), [`Error`](https://docs.oracle.com/javase/8/docs/api/java/lang/Error.html#Error-java.lang.String-java.lang.Throwable-boolean-boolean-) and [`RuntimeException`](https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html#RuntimeException-java.lang.String-java.lang.Throwable-boolean-boolean-):

```java
protected Throwable(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace)
```

Reply:
@viktorklang I think it would be nice for `ControlThrowable` to be:

```scala
class ControlThrowable(message: String) extends Throwable(message, null, false, false) {
  def this = this(null)
}
```

However, I don't know if there's a way to migrate that. Similar thought for `NoStackTrace`.